src/main/c/jnibwa.c

```diff
@@ -9,216 +9,334 @@
-		// the next two lines represent my interpretation of the SAM spec
-		// else if ( p->pos < m->pos ) kput32(m->pos+cigarRefLen(m->n_cigar, m->cigar)-p->pos, str);
-		// else kput32(m->pos-p->pos-cigarRefLen(p->n_cigar, p->cigar), str);
-		// but BWA does something else which is very odd in the case of outies,
-		// but is faithfully reproduced below
-		else {
-			// get 5' end of + reads, and 3' end of - reads
-			long p0 = p->pos;
-			if ( p->is_rev ) p0 += cigarRefLen(p->n_cigar, p->cigar) - 1;
-			long m0 = m->pos;
-			if ( m->is_rev ) m0 += cigarRefLen(m->n_cigar, m->cigar) - 1;
-			kput32(m0 - p0 + (p0 > m0 ? -1 : p0 < m0 ? 1 : 0), str);
-		}
-	}
+    if ( !which ) {
+        size_t nInts = 12; // for space planning, assume mapped, unpaired reads with 3 cigar ops and an 8 character MD
+        if ( p->flag & 0x1 ) nInts += 3; // if paired, add in enough space for mate info
+        ks_resize(str, (n*nInts+1)*sizeof(int32_t));
+        kput32(n, str);
+    }
+    int32_t flag_mapQ = p->flag;
+    if ( p->flag & 0x10000 ) flag_mapQ |= 0x100;
+    flag_mapQ = (flag_mapQ << 16) | (p->mapq & 0xff);
+    kput32(flag_mapQ, str);
+    if ( !(p->flag & 0x4) ) {
+        kput32(p->rid, str);
+        kput32(p->pos, str);
+        kput32(p->NM, str);
+        kput32(p->score, str);
+        kput32(p->sub, str);
+        int32_t nCig = p->n_cigar;
+        kput32(nCig, str);
+        uint32_t* pCig = p->cigar;
+        while ( nCig-- ) {
+            uint32_t lenOp = *pCig++;
+            // op is encoded as MIDSH in a mem_aln_t, but as MIDNSH in a BAM
+            if ( (lenOp & 0xf) > 2 ) ++lenOp;
+            kput32(lenOp, str);
+        }
+        int32_t nMD = p->n_cigar ? strlen((char*)pCig) : 0;
+        kput32(nMD, str);
+        if ( nMD )
+            kputsn((char*)pCig, (nMD+3)&~3, str);
+        int32_t nXA = p->XA ? strlen(p->XA) : 0;
+        kput32(nXA, str);
+        if ( nXA ) {
+            kputsn(p->XA, (nXA+3)&~3, str);
+        }
+    }
+    if ( (p->flag & 0x9) == 1 ) {
+        kput32(m->rid, str);
+        kput32(m->pos, str);
+        if ( (p->flag & 0x4) || p->rid != m->rid ) kput32(0, str);
+        // the next two lines represent my interpretation of the SAM spec
+        // else if ( p->pos < m->pos ) kput32(m->pos+cigarRefLen(m->n_cigar, m->cigar)-p->pos, str);
+        // else kput32(m->pos-p->pos-cigarRefLen(p->n_cigar, p->cigar), str);
+        // but BWA does something else which is very odd in the case of outies,
+        // but is faithfully reproduced below
+        else {
+            // get 5' end of + reads, and 3' end of - reads
+            long p0 = p->pos;
+            if ( p->is_rev ) p0 += cigarRefLen(p->n_cigar, p->cigar) - 1;
+            long m0 = m->pos;
+            if ( m->is_rev ) m0 += cigarRefLen(m->n_cigar, m->cigar) - 1;
+            kput32(m0 - p0 + (p0 > m0 ? -1 : p0 < m0 ? 1 : 0), str);
+        }
+    }
 }
 
 static size_t bufLen( int32_t* pBuf ) {
```

Comment:
It *is* tough to follow.  I'm adding some documentation that might make it a little clearer.